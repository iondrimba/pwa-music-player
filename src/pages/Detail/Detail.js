import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import ProgressBar from '../../components/ProgressBar/ProgressBar';
import MediaButton from '../../components/MediaButton/MediaButton';
import AlbumCover from '../../components/AlbumCover/AlbumCover';
import convertSecondsToMMss from '../../helpers/timer';
import IconButton from '../../components/IconButton/IconButton';
import { ReactComponent as PlayButton } from '../../icons/play-arrow.svg';
import { ReactComponent as PauseButton } from '../../icons/pause.svg';
import { ReactComponent as RepeatButton } from '../../icons/repeat-arrows.svg';
import { ReactComponent as LinkButton } from '../../icons/link.svg';
import './styles.scss';

class Detail extends PureComponent {
  onPlayClick = () => {
    this.props.onPlayClick(this.props.track);
  }

  onPauseClick = () => {
    this.props.onPauseClick(this.props.track);
  }

  onLinkClick = () => {
    window.open(this.props.track.permalink_url);
  }

  render() {
    return (
      <Fragment>
        <div className="detail__track" aria-live="polite" aria-atomic="false">
          <AlbumCover
            hide={!this.props.active}
            src={this.props.track.artwork_url.replace('t50x50', 't300x300')}
            alt={`album artwork from track ${this.props.track.title}`}
          />
          <div className="detail__controls">
            <div className="detail__info">
              <h3 className="title">{this.props.track.title}</h3>
              <p className="artist">{this.props.track.artist}</p>
            </div>
            <ProgressBar percent={this.props.track.percentage} />
            <div className="detail__timing">
              <time className="time">{convertSecondsToMMss(this.props.track.currentTime)}</time>
              <time className="time">{convertSecondsToMMss(this.props.track.duration / 1000)}</time>
            </div>
            <div className="detail__buttons">
              <IconButton label="repeat song" tabEnabled={this.props.active} className={`icon-button ${this.props.repeat ? 'icon-button--high-light': ''}`} onClick={this.props.onRepeatClick} icon={<RepeatButton className="icon icon--back" width={16} />} />
              <button name="previous song" tabIndex={this.props.active ? "0" : "-1"} className="prev-button" onClick={this.props.onPlayPrev}><PlayButton width={16} /></button>
              <MediaButton name="pause button" tabEnabled={this.props.active} active={this.props.track.playing && !this.props.track.paused} onClick={this.onPauseClick} icon={<PauseButton width={28} />} />
              <MediaButton name="play button" tabEnabled={this.props.active} active={!this.props.track.playing && this.props.track.paused} onClick={this.onPlayClick} icon={<PlayButton width={28} />} />
              <button name="next song button" tabIndex={this.props.active ? "0" : "-1"} className="next-button" onClick={this.props.onPlayNext}><PlayButton width={16} /></button>
              <IconButton label="song link" tabEnabled={this.props.active} className="icon-button" onClick={this.onLinkClick} icon={<LinkButton className="icon icon--back" width={16} />} />
            </div>
          </div>
        </div>
      </Fragment>
    );
  }
}

Detail.propTypes = {
  track: PropTypes.shape({
    currentTime: PropTypes.number.isRequired,
    duration: PropTypes.number,
    playing: PropTypes.bool.isRequired,
    paused: PropTypes.bool.isRequired,
    id: PropTypes.number,
    artwork_url: PropTypes.string.isRequired,
    title: PropTypes.string,
    artist: PropTypes.string,
    permalink_url: PropTypes.string
  }),
  repeat: PropTypes.bool,
  active: PropTypes.bool.isRequired,
  onRepeatClick: PropTypes.func.isRequired,
  onBackClick: PropTypes.func,
  onPauseClick: PropTypes.func.isRequired,
  onPlayClick: PropTypes.func.isRequired,
  onPlayPrev: PropTypes.func.isRequired,
  onPlayNext: PropTypes.func.isRequired,
}

export default Detail;
