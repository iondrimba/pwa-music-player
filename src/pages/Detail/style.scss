@import '../../app/styles/color';

.detail {
  background-color: $base-color;
  display: flex;
  justify-content: center;
  z-index: 2;

  .title {
    width: 300px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .title, .artist {
    padding: 0;
    margin: 0;
    text-transform: capitalize;
  }


  &__track {
    margin-top: 0;
  }

  .prev-button, .next-button {
    border: 0;
    border-radius: 50%;
    width: 50px;
    height: 50px;
    display: flex;
    justify-content: center;
    cursor: pointer;
    background-color: rgba(0, 0, 0, .09);
    transition: background-color .2s;
    position: relative;

    &:hover {
      background-color: rgba(0, 0, 0, .2);
    }

    &:active {
      background-color: rgba(0, 0, 0, .09);
    }

    svg {
      fill: $base-color;
    }

    &:before {
      content: '';
      display: block;
      position: absolute;
      opacity: .01;
      width: 100%;
      height: 100%;
      border: solid 2px $border-color;
      border-radius: 50px;
      transition: opacity .3s;
      box-shadow: 0px 0px 0px 4px rgba(123, 177, 243, 0.43);
      top: -2px;
    }
  }

  .prev-button:focus, .next-button:focus {
    outline: 0;
  }

  .prev-button:focus::before, .next-button:focus::before {
    opacity: .57;
  }


  .next-button {
    &::after {
      content: '';
      display: block;
      height: 16px;
      width: 3px;
      position: absolute;
      margin-left: 10px;
      border-radius: 6px;
      background: $base-color;
    }
  }

  .prev-button {
    &::after {
      content: '';
      display: block;
      height: 16px;
      width: 3px;
      position: absolute;
      margin-left: -10px;
      border-radius: 6px;
      background: $base-color;
    }

    svg {
      transform: scale(-1);
    }
  }

  &__controls {
    position: relative;
    display: flex;
    flex-direction: column;
  }

  &__info {
    text-align: center;
    margin-top: 50px;
    margin-bottom: 20px;
    max-width: 300px;
  }

  &__buttons {
    margin-top: 25px;
    display: flex;
    justify-content: space-between;
    position: relative;
    align-items: center;
  }

  &__timing {
    margin-top: 10px;
    display: flex;
    justify-content: space-between;
  }

  .time {
    color: $primary-font-color;
  }
}
