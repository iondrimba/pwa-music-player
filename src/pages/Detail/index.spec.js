import React from 'react';
import renderer from 'react-test-renderer';
import Detail from '.';
import { shallow } from 'enzyme';

const sampleTrack = {
  id: 1,
  'artwork_url': 'artwork_url',
  title: 'title',
  artist: 'artist',
  percentage: 1,
  currentTime: 360,
  duration: 450,
  playing: false,
  paused: false,
  permalink_url: 'permalink_url',
};

const detail = ({
  track,
  onPlayClick = () => { },
  onRepeatClick = () => { },
  onPlayNext = () => { },
  onBackClick = () => { },
  onPlayPrev = () => { },
  onPauseClick = () => { } },
) => {
  return <Detail track={track}
    active={true}
    onRepeatClick={onRepeatClick}
    onPlayClick={onPlayClick}
    onPlayNext={onPlayNext}
    onBackClick={onBackClick}
    onPlayPrev={onPlayPrev}
    onPauseClick={onPauseClick} />
};

describe('Detail', () => {
  it('renders without errors', () => {
    const component = renderer.create(detail({ track: sampleTrack }));

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('onPlayClick', () => {
    it('calls props.onPlayClick', () => {
      const onPlayClick = jest.fn();
      const component = shallow(detail({ track: sampleTrack, onPlayClick }));

      component.instance().onPlayClick();

      expect(onPlayClick).toBeCalledWith(component.instance().props.track);
    });
  });

  describe('onPauseClick', () => {
    it('calls props.onPauseClick', () => {
      const onPauseClick = jest.fn();
      const component = shallow(detail({ track: sampleTrack, onPauseClick }));

      component.instance().onPauseClick();

      expect(onPauseClick).toBeCalledWith(component.instance().props.track);
    });
  });


  describe('onLinkClick', () => {
    it('calls window.open', () => {
      window.open = jest.fn();
      const onLinkClick = jest.fn();
      const component = shallow(detail({ track: sampleTrack, onLinkClick }));

      component.instance().onLinkClick();

      expect(window.open).toBeCalledWith(component.instance().props.track.permalink_url);
    });
  });
});
