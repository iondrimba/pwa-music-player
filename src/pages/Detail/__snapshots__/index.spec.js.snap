// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail renders without errors 1`] = `
<div
  aria-atomic="false"
  aria-live="polite"
  className="detail__track"
>
  <div
    className="album-cover"
  >
    <svg
      className="album-cover__icon"
    >
      picture.svg
    </svg>
    <div
      className="album-cover__loader"
    />
    <img
      alt="album artwork from track title"
      className="album-cover__image"
      crossOrigin=""
      onLoad={[Function]}
    />
  </div>
  <div
    className="detail__controls"
  >
    <div
      className="detail__info"
    >
      <h3
        className="title"
      >
        title
      </h3>
      <p
        className="artist"
      >
        artist
      </p>
    </div>
    <div
      className="progress-bar"
    >
      <div
        className="progress-bar__mask"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <div
          className="progress-bar__progress"
        />
      </div>
      <div
        className="progress-bar__bg"
      />
    </div>
    <div
      className="detail__timing"
    >
      <time
        className="time"
      >
        06:00
      </time>
      <time
        className="time"
      >
        00:00
      </time>
    </div>
    <div
      className="detail__buttons"
    >
      <button
        aria-label="repeat song"
        className="icon-button icon-button "
        name="repeat song"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        <svg
          className="icon icon--back"
          width={16}
        >
          repeat-arrows.svg
        </svg>
      </button>
      <button
        className="prev-button"
        name="previous song"
        onClick={[Function]}
        tabIndex="0"
      >
        <svg
          width={16}
        >
          play-arrow.svg
        </svg>
      </button>
      <button
        aria-label="pause button"
        className=" media-button "
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        <svg
          width={28}
        >
          pause.svg
        </svg>
      </button>
      <button
        aria-label="play button"
        className=" media-button "
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        <svg
          width={28}
        >
          play-arrow.svg
        </svg>
      </button>
      <button
        className="next-button"
        name="next song button"
        onClick={[Function]}
        tabIndex="0"
      >
        <svg
          width={16}
        >
          play-arrow.svg
        </svg>
      </button>
      <button
        aria-label="song link"
        className="icon-button icon-button"
        name="song link"
        onClick={[Function]}
        tabIndex="0"
        type="button"
      >
        <svg
          className="icon icon--back"
          width={16}
        >
          link.svg
        </svg>
      </button>
    </div>
  </div>
</div>
`;
